### PR TITLE
Add core API classes to support BQ Pushdown for aggregations

### PR DESCRIPTION
**Pull request by DJSagarAhire**
This PR adds core API classes to support BQ pushdown for the group by and deduplicate aggregations using the declarative plugins framework.

### CODE REVIEW

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright © 2021 Cask Data, Inc.
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java]: 2022

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Specifies how an aggregation should be executed.
+ */
+@Beta
+public abstract class AggregationDefinition {
+  private final List<Expression> groupByFields;
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java]: Could you name it "groupByExpressions/selectExpressions"? It's consistent with types

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    groupByFields = Collections.emptyList();
+    selectedFields = Collections.emptyList();
+  }
+
+  protected AggregationDefinition(List<Expression> groupByFields, List<Expression> selectedFields) {
+    this.groupByFields = Collections.unmodifiableList(groupByFields);
+    this.selectedFields = Collections.unmodifiableList(selectedFields);
+  }
+
+  public List<Expression> getGroupByFields() {
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java]: Please add javadocs to all public / protected methods

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.etl.api.aggregation;
+
+import io.cdap.cdap.etl.api.relational.Expression;
+import javafx.util.Pair;
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: Don't use javafx

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Specifies how a deduplicate operation should be executed.
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:28]: Please put a really good description here  and/or in the "Relation". It's non-standard functionality

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+ */
+public class DeduplicateAggregationDefinition extends AggregationDefinition {
+  List<Pair<Expression, FilterFunction>> filterFields;
+
+  private DeduplicateAggregationDefinition(List<Expression> groupByFields,
+                                           List<Expression> selectFields,
+                                           List<Pair<Expression, FilterFunction>> filterFields) {
```
Comment (fernst) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: Maybe this `Pair<Expression, FilterFunction>` should be it's own class?

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinField.java [hunk 7]
```diff
@@ -63,7 +74,7 @@ public int hashCode() {
 
   @Override
   public String toString() {
-    return "JoinField{" +
+    return "Field{" +
```
Comment (fernst) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinField.java]: The class name is still the same, right? This should still be called `JoinField`

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (!(o instanceof StringExpression)) {
+      return false;
+    }
+    StringExpression that = (StringExpression) o;
+    return getExpression().equals(that.getExpression());
```
Comment (fernst) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java]: This equals method doesn't have null checks. 

I would check if `o` is null and use `Objects.equals()` to perform the equality comparison.

**Review by fernst — COMMENTED** [review]
Left some comments/observations

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java (hunk 1 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java]: I started writing that class in 2021... 😅 Fixed.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java (hunk 2 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java]: Done.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java (hunk 3 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java]: Done, let me know if the javadoc looks good and if I missed anything.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java (hunk 4 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: Changed to `AbstractMap.SimpleEntry`.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java (hunk 5 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:28]: Added, let me know if it looks good.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java (hunk 6 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: Changed to `AbstractMap.SimpleEntry`.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinField.java (hunk 7 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinField.java]: Done.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java (hunk 8 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java]: Fixed this, although since the class doesn't have a default constructor I don't quite see how `getExpression()` will be null.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java (hunk 5 above)
Comment (tivv) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:28]: Yes. Please add some links to the respective getters / classes (e.g. where groupByExpressions/selectExpressions/filterExpressions are used)

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+public abstract class AggregationDefinition {
+  private final List<Expression> groupByExpressions;
+  private final List<Expression> selectedExpressions;
+
+  /**
+   * Creates a default {@link AggregationDefinition} object with empty lists for group by and select expressions.
+   */
+  protected AggregationDefinition() {
+    groupByExpressions = Collections.emptyList();
+    selectedExpressions = Collections.emptyList();
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java]: Use "selectExpression". Or even "resultExpressions"  as "select" is too tied to SQL and in fact it would define expressions that form the result Relation

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    super(groupByExpressions, selectExpressions);
+    this.filterExpressions = filterExpressions;
+  }
+
+  /**
+   * Get the list of expressions using which a single record will be selected from a group of records for deduplication.
+   * Each entry in the list is defined as a pair of an {@link Expression} and {@link FilterFunction}, specifying
+   * the function to be executed on the expression.
+   * @return A {@link List} of {@link SimpleEntry} objects, each of which is a mapping between an {@link Expression}
+   * and a {@link FilterFunction}.
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:53]: I'd not say it's a mapping. It's not used as a map, rather as two parameters to select one row out of the group

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java [hunk 11]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Specifies how a deduplicate operation should be executed.
+ * A deduplication is an operation that takes in a set of rows and for each group of rows for which the values for
+ * certain group by fields are the same, outputs a single row.
+ * This single row can be determined by performing an operation on the set of rows which returns a single row.
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:31]: Could you expand on this? Basically filterExpressions define a composite key and a row with the specified (minimum or maximum according to the function for the column) key is selected

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java [hunk 12]
```diff
@@ -0,0 +1,129 @@
...
+    private List<Expression> selectExpressions;
+    List<SimpleEntry<Expression, FilterFunction>> filterExpressions;
+
+    public Builder() {
+      groupByExpressions = Collections.emptyList();
+      selectExpressions = Collections.emptyList();
+      filterExpressions = new ArrayList<>();
+    }
+
+    public Builder dedupOn(List<Expression> dedupExpressions) {
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:148]: Please also add javadocs to the builder. It helps when working from IDE and building the definition

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/GroupByAggregationDefinition.java [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+  public static class Builder {
+    private List<Expression> groupByExpressions;
+    private List<Expression> selectExpressions;
+
+    public Builder() {
+      groupByExpressions = Collections.emptyList();
+      selectExpressions = Collections.emptyList();
+    }
+
+    public Builder groupBy(List<Expression> groupByExpressions) {
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/GroupByAggregationDefinition.java:60]: Please also add javadocs to the builder. It helps when working from IDE and building the definition

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinField.java [hunk 14]
```diff
@@ -26,22 +25,34 @@
  * The name of a field and an optional alias to rename it to.
  */
 @Beta
-public class JoinField extends Field {
+public class JoinField {
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinField.java:28]: Please don't change existing API. It's already public, so it's better not to change it unless really needed.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+package io.cdap.cdap.etl.api.relational;
+
+import java.util.Objects;
+
+/**
+ * A default implementation of Expression that simply stores the expression as a string.
+ * The string is assumed to be ANSI SQL compliant.
+ * It is the responsibility of the factory creating StringExpression objects to enforce correctness of the expression.
+ */
+public class StringExpression implements Expression {
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java]: This are implementations. It's better not to put them into API as this creates additional contract hard to refactor / extend.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpressionFactoryType.java [hunk 16]
```diff
@@ -14,24 +14,14 @@
  * the License.
  */
 
-package io.cdap.cdap.etl.api.aggregation;
+package io.cdap.cdap.etl.api.relational;
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpressionFactoryType.java]: This renames looks weird. Could you please check why git picks this set of changes as renames? It's hard to follow

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinField.java (hunk 14 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/join/JoinField.java:28]: This is not in any CDAP version, this is just.a revert of [an earlier BQ Pushdown PR](https://github.com/cdapio/cdap/pull/13746) which added the Fields classes, which are no longer required.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpressionFactoryType.java (hunk 16 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpressionFactoryType.java]: Not sure why git picks it as a rename. I've deleted the `GroupByFields` class since we're not using Field classes any more and added the `StringExpressionFactoryType` class (technically it's a rename of `StringExpressionFactory` since it's actually an implementation of `ExpressionFactoryType`), but oddly the diff treats it as a file rename rather than two separate file delete and file addition operations.

If anything the diff should have treated `StringExpressionFactoryType` as a rename of `StringExpressionFactory` and then flagged the new `StringExpressionFactory` class as a new file, as that was the actual intent of the PR.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java (hunk 15 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java]: Do you suggest I should move this class inside the factory class instead?

Comment (tivv) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java]: It should be in the google-cloud plugin, not in the cdap-etl-api

Comment (tivv) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java]: Overall this is convenient implementation, but not part of API.

Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java]: So in this case `StringExpressionFactory` will have to move to google-cloud as well, correct?

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java (hunk 11 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:31]: A single filterExpressions pair defines a single key and a function (MIN or MAX). If this function returns more than one row, then the next filterExpressions pair is used and so on.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java (hunk 9 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java]: Changed to `selectExpression`. I believe we should stick to SQL nomenclature as everyone is familiar with it.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java (hunk 10 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:53]: Fixed.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java (hunk 12 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:148]: Added.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/GroupByAggregationDefinition.java (hunk 13 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/GroupByAggregationDefinition.java:60]: Added.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java (hunk 11 above)
Comment (tivv) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:31]: I mean put better explanation in the javadoc, so that a person who does not know how dedup works would understand it

Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java:31]: I know, I did that too :)

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java (hunk 15 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/relational/StringExpression.java]: Removed, adding it in google-cloud now.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java [hunk 17]
```diff
@@ -0,0 +1,63 @@
...
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.etl.api.Field;
+
+import java.util.Collections;
+import java.util.List;
+
+/**
+ * Specifies how an aggregation should be executed.
+ */
+@Beta
```
Comment (sanjanasandeep) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/AggregationDefinition.java:28]: until when do we keep this annotation , i.e; until when is it in beta?

**Review by sanjanasandeep — APPROVED** [review]
LGTM

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java [hunk 18]
```diff
@@ -0,0 +1,176 @@
...
+    /**
+     * Sets the list of filter expressions to the specified list of filter functions and filter expressions.
+     * The filter function is applied to the respective filter expression for all the records for which the values
+     * of the dedup expressions specified in <code>dedupOn</code> is the same, and the resultant row is extracted
+     * by the dedup operation.
+     * @param filterExpressions a {@link List} of {@link SimpleEntry} pairs of
+     *                          an {@link Expression} and a {@link FilterFunction}.
+     * @return a {@link Builder} with the currently built {@link DeduplicateAggregationDefinition}.
+     */
+    public Builder filterDuplicatesBy(List<SimpleEntry<Expression, FilterFunction>> filterExpressions) {
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: please use interface instead of implementation as the parameter type (e.g. `List<? extends Map.Entry<Expression, FilterFunction>>`)

**Review by tivv — COMMENTED** [review]
One small comment left (about Map.Entry), otherwise LGTM

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java (hunk 18 above)
Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: This is done, PTAL.

File: cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java [hunk 19]
```diff
@@ -0,0 +1,177 @@
...
+     * Sets the list of filter expressions to the specified list of filter functions and filter expressions.
+     * The filter function is applied to the respective filter expression for all the records for which the values
+     * of the dedup expressions specified in <code>dedupOn</code> is the same, and the resultant row is extracted
+     * by the dedup operation.
+     * @param filterExpressions a {@link List} of {@link SimpleEntry} pairs of
+     *                          an {@link Expression} and a {@link FilterFunction}.
+     * @return a {@link Builder} with the currently built {@link DeduplicateAggregationDefinition}.
+     */
+    public Builder filterDuplicatesBy(List<? extends Map.Entry<Expression, FilterFunction>> filterExpressions) {
+      this.filterExpressions = (List<SimpleEntry<Expression, FilterFunction>>) filterExpressions;
```
Comment (tivv) [inline cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: Why do you need this cast? You should be fine with `filterExpression` fields being list of `Map.Entry` as well, should not you?

Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: I need to create a new entry to add to the list in the two-argument version of `filterDuplicatesBy()` which would be a `SimpleEntry`. I can't add `SimpleEntry` elements to a `? extends Map.Entry` list due to possible list pollution, so we need to keep the internal list as a `SimpleEntry` list.

Comment (tivv) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: OK, what you need to do is to have `List<Map.Entry` as a field and use clear / addAll instead of taking directly user list. But you can't be sure that user passes you `SimpleEntry` objects. Or introduce your own tuple

Comment (DJSagarAhire) [reply cdap-app-templates/cdap-etl/cdap-etl-api/src/main/java/io/cdap/cdap/etl/api/aggregation/DeduplicateAggregationDefinition.java]: Created a new tuple class.

**Review by fernst — APPROVED** [review]
I see all comments have been addresses. LGTM.
